### ISSUE TITLE
There are more than 3 contributors in the project

### ISSUE THREAD

**Issue by pedestrianlove** [ISS-SOA/soa2024-codepraise#10]
![image](https://github.com/user-attachments/assets/9b111332-b04d-4e1f-8b39-b1c6e69d8ea4)

![image](https://github.com/user-attachments/assets/ca733bb9-4a09-4f66-ac33-6931aea0c484)

Would it be possible to map the git blame record according to the email address instead of name?

**Comment by soumyaray** [comment]
Yes I'd like to resolve this issue as well, because it happens to many groups. I currently combine blame data from two users with the same _email_.  I believe the issue you are facing is because there are commits from five seemingly different users (i.e., unique emails):

```
Wuyuwin -> winwu93@gmail.com
J.S.LI -> jsli@linux.com
harvey -> sideman8787@gmail.com
pedestrianlove -> 37164678+pedestrianlove@users.noreply.github.com
Wuyuwin -> 121348598+Wuyuwin@users.noreply.github.com
```

This happens when users change their email address over different commits. Emails of the form `###+abc@users.noreply.github.com` indicate the git user's email was kept private on Github and so it is not listed on a commit. Any suggestions on how to resolve the above?

For now, I can think of just combining information where username and/or emails are the same.

**Comment by pedestrianlove** [comment]
I tried to look for solutions on Google, and it seems to be possible to extract the unique identity of a member/contributor by the GitHub username:

- For the following 2 cases, fetch and use the GitHub username as the unique identifier:
  - For the pure email part: use the GitHub API to map to the GitHub username: https://stackoverflow.com/questions/44888187/get-github-username-through-primary-email
  - For the `hash+username@users.noreply.github.com` part, we can parse the string according to the format listed here and get the username: https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address#about-commit-email-addresses
- If there are any other outlier cases or parsing error, then we can give it something like `unknown-{detected name}-{detected email}` as the fallback unique identifier. 

I haven't yet tested the GitHub API, so not sure if it would work as expected though.

**Comment by pedestrianlove** [comment]
OK, I've just tested the API with the gh cli, it won't work if you set your email as private:
![image](https://github.com/user-attachments/assets/ca8c2154-491f-4894-85b7-d3b58e8b77fd)

Update:
You can get the username via the the commit hash: https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#get-a-commit
![image](https://github.com/user-attachments/assets/445499eb-0cd0-403d-b7d5-829ba57f89a4)

I think it should also be possible to cache the correspondence of identifier information(info in git blame) to the actual GitHub username in a  hash or even database table to further improve the performance.

**Comment by soumyaray** [comment]
Thanks much for checking it out – let me think up some solutions...

**Comment by soumyaray** [comment]
I took a deeper dive into this. It seems I can also get committer info from the local git repo, but it does not provide the non-private email:

```
$ git show --pretty=format:"%an <%ae>" --no-patch cd14d484

Wuyuwin <121348598+Wuyuwin@users.noreply.github.com>
```

And neither does Github's API:

```
$ gh api /repos/SOA-04/leaf-api/commits/cd14d484

{
  "sha": "cd14d484501eb2ad639b97eb969ed842582bb16b",
  "node_id": "C_kwDOM5HeqtoAKGNkMTRkNDg0NTAxZWIyYWQ2MzliOTdlYjk2OWVkODQyNTgyYmIxNmI",
  "commit": {
    "author": {
      "name": "Wuyuwin",
      "email": "121348598+Wuyuwin@users.noreply.github.com",
      "date": "2024-10-27T00:51:40Z"
    },
    "committer": {
      "name": "GitHub",
      "email": "noreply@github.com",
      "date": "2024-10-27T00:51:40Z"
    },
    "message": "Update 001_create_locations_table.rb",
    "tree": {
      "sha": "18273081e0a2f96a01d0b0a0163024f44d2fe9b4",
      "url": "https://api.github.com/repos/SOA-04/leaf-api/git/trees/18273081e0a2f96a01d0b0a0163024f44d2fe9b4"
    },
```

Also complicating the issue for your particular repo is that the same user changes their username (e.g., `Wuyuwin` and `wuyuwin`) while also changing their email. Thus, without intimate knowledge of the committers, I don't see how CodePraise can automatically resolve the situation.

One possible solution is to ask users to manually merge committers – but here I'm not sure if this should be done on a per-user or per-project basis. Welcome to make more suggestions along this line.

Regrettably, unless this issue becomes a showstopper problem, I will leave it unresolved for now until an agreeable long-term solution presents itself.

**Comment by pedestrianlove** [comment]
> I took a deeper dive into this. It seems I can also get committer info from the local git repo, but it does not provide the non-private email:
> 
> 
> 
> ```
> 
> $ git show --pretty=format:"%an <%ae>" --no-patch cd14d484
> 
> 
> 
> Wuyuwin <121348598+Wuyuwin@users.noreply.github.com>
> 
> ```
> 
> 
> 
> And neither does Github's API:
> 
> 
> 
> ```
> 
> $ gh api /repos/SOA-04/leaf-api/commits/cd14d484
> 
> 
> 
> {
> 
>   "sha": "cd14d484501eb2ad639b97eb969ed842582bb16b",
> 
>   "node_id": "C_kwDOM5HeqtoAKGNkMTRkNDg0NTAxZWIyYWQ2MzliOTdlYjk2OWVkODQyNTgyYmIxNmI",
> 
>   "commit": {
> 
>     "author": {
> 
>       "name": "Wuyuwin",
> 
>       "email": "121348598+Wuyuwin@users.noreply.github.com",
> 
>       "date": "2024-10-27T00:51:40Z"
> 
>     },
> 
>     "committer": {
> 
>       "name": "GitHub",
> 
>       "email": "noreply@github.com",
> 
>       "date": "2024-10-27T00:51:40Z"
> 
>     },
> 
>     "message": "Update 001_create_locations_table.rb",
> 
>     "tree": {
> 
>       "sha": "18273081e0a2f96a01d0b0a0163024f44d2fe9b4",
> 
>       "url": "https://api.github.com/repos/SOA-04/leaf-api/git/trees/18273081e0a2f96a01d0b0a0163024f44d2fe9b4"
> 
>     },
> 
> ```
> 
> 
> 
> Also complicating the issue for your particular repo is that the same user changes their username (e.g., `Wuyuwin` and `wuyuwin`) while also changing their email. Thus, without intimate knowledge of the committers, I don't see how CodePraise can automatically resolve the situation.
> 
> 
> 
> One possible solution is to ask users to manually merge committers – but here I'm not sure if this should be done on a per-user or per-project basis. Welcome to make more suggestions along this line.
> 
> 
> 
> Regrettably, unless this issue becomes a showstopper problem, I will leave it unresolved for now until an agreeable long-term solution presents itself.

My point is, it would probably be a good idea to make the unique identifier of a contributor based solely on GitHub username. You can check out the `login` attribute in the GitHub API response.

Another possible workaround may be just write another interactive cli program to allow students to remap their commit info into their actual email/username on a regular basis.

But anyway, I'm just firing up this issue to break the ice 😂, so it's not really a serious issue to me whatsoever 😊.

**Comment by soumyaray** [comment]
I think it would be difficult to correlate github login with git committer attributes because there is not mechanism tying the two together. Also consider an early project where some of the committers have not created a proper github profile yet.

But your suggestion about asking students to define contributor profiles gives me a very workable idea: we could ask maintainers to leave a CONTRIBUTORS.md/json or AUTHORS.md/json file to indicate how to identify committers based on emails/usernames, etc.  I like this solution because it doesn't require any front-end complications and the mapping is adjustable in the repo itself for any platform doing similar analysis.

I think I will move forward with that idea, but not sure if it will find its way into the class version of the project – I don't want to complicate the code beyond what I've explained in class.  But I'm very seriously considering making a more public version of CodePraise that implements this and other often asked for features! Thanks for engaging on this!

**Comment by soumyaray** [comment]
@pedestrianlove I've semi-solved the issue by cross-checking all contributors and identifying a unique contributor when there are multiple possibilities.  However, in the case of your project, its a bit more complicated because different subfolders will return different names for you or your teammates (pedestrianlove / J.S.Li).  So the results will be a bit glitchy for your specific leaf-api project. A longer term solution will be to create a master record of unique contributors.  That might have to wait till after the semester!  I don't think this will be an issue for most OSS projects.  I'll close this issue soon unless there is some pending work to be done.

**Comment by pedestrianlove** [comment]
> @pedestrianlove I've semi-solved the issue by cross-checking all contributors and identifying a unique contributor when there are multiple possibilities. However, in the case of your project, its a bit more complicated because different subfolders will return different names for you or your teammates (pedestrianlove / J.S.Li). So the results will be a bit glitchy for your specific leaf-api project. A longer term solution will be to create a master record of unique contributors. That might have to wait till after the semester! I don't think this will be an issue for most OSS projects. I'll close this issue soon unless there is some pending work to be done.

Cool! Thanks🥳🥳🥳.
